### PR TITLE
Rework sorting and add cell path and closure comparators to `sort-by`

### LINKED ISSUE #8322: Support sub-record keys and closures in `group-by`/`sort-by`
### Related problem

It is not easy to group/sort data with sub-keys.

Consider the following data structure storing Twitter tweets.

```jsonc
[
  {
    "author": { 
      "id": 1,
      "username": "mon-jai"
	},
    "created": 1677877558
    "caption": "Hello world",
    "comments": [{}, {}, {}]
  },
  // ...
]
```

In today's Nushell,

What if I want to group tweets by author?

```
$tweets | each {|$it| insert author_username { $it | get author.username } } | group-by author_username
```

What if I want to sort tweets by comments count?

```
$tweets | each {|$it| insert comments_count { $it | get comments | length } } | group-by comments_count
```

The solution requires us to process the data beforehand and not quite straight forward.

### Describe the solution you'd like

By supporting sub-record keys and closures in `group-by`/`sort-by`, we can process the data more conveniently.

```
$tweets | group-by author.username
```

```
$tweets | sort-by {|it| $it | get comments | length }
```

### Describe alternatives you've considered

_No response_

### Additional context and details

_No response_

### LINKED ISSUE #12535: Support `sort-by` with nested columns
### Related problem

My data has this shape:
```json
[{"metadata": {"creationTimestamp":"2024-04-15T12:11:00Z"}}]
```

### Describe the solution you'd like

I would like to `sort-by metadata.creationTimestamp`, like I can do `get metadata.creationTimestamp`.

### Describe alternatives you've considered

I could revert to `jq`, but I would prefer a solution with nushell's native tools.

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by 132ikl**
# Description

Closes #12535
Implements sort-by functionality of #8322
Fixes sort-by part of #8667

This PR does two main things: add a new cell path and closure parameter to `sort-by`, and attempt to make Nushell's sorting behavior well-defined.

## `sort-by` features

The `columns` parameter is replaced with a `comparator` parameter, which can be a cell path or a closure. Examples are from docs PR.

1. Cell paths

   The basic interactive usage of `sort-by` is the same. For example, `ls | sort-by modified` still works the same as before. It is not quite a drop-in replacement, see [behavior changes](#behavior-changes).
   
   Here's an example of how the cell path comparator might be useful:
   
   ```nu
   > let cities = [
       {name: 'New York', info: { established: 1624, population: 18_819_000 } }
       {name: 'Kyoto', info: { established: 794, population: 37_468_000 } }
       {name: 'São Paulo', info: { established: 1554, population: 21_650_000 } }
   ]
   > $cities | sort-by info.established
   ╭───┬───────────┬────────────────────────────╮
   │ # │   name    │            info            │
   ├───┼───────────┼────────────────────────────┤
   │ 0 │ Kyoto     │ ╭─────────────┬──────────╮ │
   │   │           │ │ established │ 794      │ │
   │   │           │ │ population  │ 37468000 │ │
   │   │           │ ╰─────────────┴──────────╯ │
   │ 1 │ São Paulo │ ╭─────────────┬──────────╮ │
   │   │           │ │ established │ 1554     │ │
   │   │           │ │ population  │ 21650000 │ │
   │   │           │ ╰─────────────┴──────────╯ │
   │ 2 │ New York  │ ╭─────────────┬──────────╮ │
   │   │           │ │ established │ 1624     │ │
   │   │           │ │ population  │ 18819000 │ │
   │   │           │ ╰─────────────┴──────────╯ │
   ╰───┴───────────┴────────────────────────────╯
   ```

2. Key closures

   You can supply a closure which will transform each value into a sorting key (without changing the underlying data). Here's an example of a key closure, where we want to sort a list of assignments by their average grade:

   ```nu
   > let assignments = [
       {name: 'Homework 1', grades: [97 89 86 92 89] }
       {name: 'Homework 2', grades: [91 100 60 82 91] }
       {name: 'Exam 1', grades: [78 88 78 53 90] }
       {name: 'Project', grades: [92 81 82 84 83] }
   ]
   > $assignments | sort-by { get grades | math avg }
   ╭───┬────────────┬───────────────────────╮
   │ # │    name    │        grades         │
   ├───┼────────────┼───────────────────────┤
   │ 0 │ Exam 1     │ [78, 88, 78, 53, 90]  │
   │ 1 │ Project    │ [92, 81, 82, 84, 83]  │
   │ 2 │ Homework 2 │ [91, 100, 60, 82, 91] │
   │ 3 │ Homework 1 │ [97, 89, 86, 92, 89]  │
   ╰───┴────────────┴───────────────────────╯
   ```

3. Custom sort closure

   The `--custom`, or `-c`, flag will tell `sort-by` to interpret closures as custom sort closures. A custom sort closure has two parameters, and returns a boolean. The closure should return `true` if the first parameter comes _before_ the second parameter in the sort order.
   
   For a simple example, we could rewrite a cell path sort as a custom sort (see [here](https://github.com/nushell/nushell.github.io/pull/1568/files#diff-a7a233e66a361d8665caf3887eb71d4288000001f401670c72b95cc23a948e86R231) for a more complex example):
   
   ```nu
   > ls | sort-by -c {|a, b| $a.size < $b.size }
   ╭───┬─────────────────────┬──────┬──────────┬────────────────╮
   │ # │        name         │ type │   size   │    modified    │
   ├───┼─────────────────────┼──────┼──────────┼────────────────┤
   │ 0 │ my-secret-plans.txt │ file │    100 B │ 10 minutes ago │
   │ 1 │ shopping_list.txt   │ file │    100 B │ 2 months ago   │
   │ 2 │ myscript.nu         │ file │  1.1 KiB │ 2 weeks ago    │
   │ 3 │ bigfile.img         │ file │ 10.0 MiB │ 3 weeks ago    │
   ╰───┴─────────────────────┴──────┴──────────┴────────────────╯
   ```
   

## Making sort more consistent

I think it's important for something as essential as `sort` to have well-defined semantics. This PR contains some changes to try to make the behavior of `sort` and `sort-by` consistent. In addition, after working with the internals of sorting code, I have a much deeper understanding of all of the edge cases. Here is my attempt to try to better define some of the semantics of sorting (if you are just interested in changes, skip to "User-Facing changes")

- `sort`, `sort -v`, and `sort-by` now all work the same. Each individual sort implementation has been refactored into two functions in `sort_utils.rs`: `sort`, and `sort_by`. These can also be used in other parts of Nushell where values need to be sorted.
  - `sort` and `sort-by` used to handle `-i` and `-n` differently.
    - `sort -n` would consider all values which can't be coerced into a string to be equal
    - `sort-by -i` and `sort-by -n` would only work if all values were strings
    - In this PR, insensitive sort only affects comparison between strings, and natural sort only applies to numbers and strings (see below).
- (not a change) Before and after this PR, `sort` and `sort-by` support sorting mixed types. There was a lot of discussion about potentially making `sort` and `sort-by` only work on lists of homogeneous types, but the general consensus was that `sort` should not error just because its input contains incompatible types.
  - In order to try to make working with data containing `null` values easier, I changed the PartialOrd order to sort `Nothing` values to the end of a list, regardless of what other types the list contains. Before, `null` would be sorted before `Binary`, `CellPath`, and `Custom` values.
  - (not a change) When sorted, lists of mixed types will contain sorted values of each type in order, for the most part
    - (not a change) For example, `[0x[1] (date now) "a" ("yesterday" | into datetime) "b" 0x[0]]` will be sorted as `["a", "b", a day ago, now, [0], [1]]`, where sorted strings appear first, then sorted datetimes, etc.
    - (not a change) The exception to this is `Int`s and `Float`s, which will intermix, `Strings` and `Glob`s, which will intermix, and `None` as described above. Additionally, natural sort will intermix strings with ints and floats (see below).
- Natural sort no longer coerce all inputs to strings.
  - I did originally make natural only apply to strings, but @fdncred pointed out that the previous behavior also allowed you to sort numeric strings with numbers. This seems like a useful feature if we are trying to support sorting with mixed types, so I settled on coercing only numbers (int, float). This can be reverted if people don't like it.
    - Here is an example of this behavior in action, which is the same before and after this PR:
      ```nushell
      $ [1 "4" 3 "2"] | sort --natural
      ╭───┬───╮
      │ 0 │ 1 │
      │ 1 │ 2 │
      │ 2 │ 3 │
      │ 3 │ 4 │
      ╰───┴───╯
      ```

# User-Facing Changes

## New features

- Replaces the `columns` string parameter of `sort-by` with a cell path or a closure. 
  - The cell path parameter works exactly as you would expect
  - By default, the `closure` parameter acts as a "key sort"; that is, each element is transformed by the closure into a sorting key
  - With the `--custom` (`-c`) parameter, you can define a comparison function for completely custom sorting order.

## Behavior changes

<details>
<summary><code>sort -v</code> does not coerce record values to strings</summary>

This was a bit of a surprising behavior, and is now unified with the behavior of `sort` and `sort-by`. Here's an example where you can observe the values being implicitly coerced into strings for sorting, as they are sorted like strings rather than numbers:

Old behavior:

```nushell
$ {foo: 9 bar: 10} | sort -v
╭─────┬────╮
│ bar │ 10 │
│ foo │ 9  │
╰─────┴────╯
```

New behavior:

```nushell
$ {foo: 9 bar: 10} | sort -v
╭─────┬────╮
│ foo │ 9  │
│ bar │ 10 │
╰─────┴────╯
```

</details>

<details>
<summary>Changed <code>sort-by</code> parameters from <code>string</code> to <code>cell-path</code> or <code>closure</code>. Typical interactive usage is the same as before, but if passing a variable to <code>sort-by</code> it must be a cell path (or closure), not a string</summary>

Old behavior:

```nushell
$ let sort = "modified"
$ ls | sort-by $sort
╭───┬──────┬──────┬──────┬────────────────╮
│ # │ name │ type │ size │    modified    │
├───┼──────┼──────┼──────┼────────────────┤
│ 0 │ foo  │ file │  0 B │ 10 hours ago   │
│ 1 │ bar  │ file │  0 B │ 35 seconds ago │
╰───┴──────┴──────┴──────┴────────────────╯
```

New behavior:

```nushell
$ let sort = "modified"
$ ls | sort-by $sort
Error: nu::shell::type_mismatch

  × Type mismatch.
   ╭─[entry #10:1:14]
 1 │ ls | sort-by $sort
   ·              ──┬──
   ·                ╰── Cannot sort using a value which is not a cell path or closure
   ╰────
$ let sort = $."modified"
$ ls | sort-by $sort
╭───┬──────┬──────┬──────┬───────────────╮
│ # │ name │ type │ size │   modified    │
├───┼──────┼──────┼──────┼───────────────┤
│ 0 │ foo  │ file │  0 B │ 10 hours ago  │
│ 1 │ bar  │ file │  0 B │ 2 minutes ago │
╰───┴──────┴──────┴──────┴───────────────╯
```
</details>

<details>
<summary>Insensitve and natural sorting behavior reworked</summary>

Previously, the `-i` and `-n` worked differently for `sort` and `sort-by` (see "Making sort more consistent"). Here are examples of how these options result in different sorts now:

1. `sort -n`
    - Old behavior (types other than numbers, strings, dates, and binary sorted incorrectly)
      ```nushell
      $ [2sec 1sec] | sort -n
      ╭───┬──────╮
      │ 0 │ 2sec │
      │ 1 │ 1sec │
      ╰───┴──────╯
      ```
    - New behavior
      ```nushell
      $ [2sec 1sec] | sort -n
      ╭───┬──────╮
      │ 0 │ 1sec │
      │ 1 │ 2sec │
      ╰───┴──────╯
      ```
    
2. `sort-by -i`
   - Old behavior (uppercase words appear before lowercase words as they would in a typical sort, indicating this is not actually an insensitive sort)
     ```nushell
     $ ["BAR" "bar" "foo" 2 "FOO" 1] | wrap a | sort-by -i a
     ╭───┬─────╮
     │ # │  a  │
     ├───┼─────┤
     │ 0 │   1 │
     │ 1 │   2 │
     │ 2 │ BAR │
     │ 3 │ FOO │
     │ 4 │ bar │
     │ 5 │ foo │
     ╰───┴─────╯
     ```
   - New behavior (strings are sorted stably, indicating this is an insensitive sort)
     ```nushell
     $ ["BAR" "bar" "foo" 2 "FOO" 1] | wrap a | sort-by -i a
     ╭───┬─────╮
     │ # │  a  │
     ├───┼─────┤
     │ 0 │   1 │
     │ 1 │   2 │
     │ 2 │ BAR │
     │ 3 │ bar │
     │ 4 │ foo │
     │ 5 │ FOO │
     ╰───┴─────╯
     ```

3. `sort-by -n`
   - Old behavior (natural sort does not work when data contains non-string values)
     ```nushell
     $ ["10" 8 "9"] | wrap a | sort-by -n a
     ╭───┬────╮
     │ # │ a  │
     ├───┼────┤
     │ 0 │  8 │
     │ 1 │ 10 │
     │ 2 │ 9  │
     ╰───┴────╯
     ```
   - New behavior
     ```nushell
     $ ["10" 8 "9"] | wrap a | sort-by -n a
     ╭───┬────╮
     │ # │ a  │
     ├───┼────┤
     │ 0 │  8 │
     │ 1 │ 9  │
     │ 2 │ 10 │
     ╰───┴────╯
     ```

</details>

<details>
<summary>
Sorting a list of non-record values with a non-existent column/path now errors instead of sorting the values directly (<code>sort</code> should be used for this, not <code>sort-by</code>)
</summary>

Old behavior:

```nushell
$ [2 1] | sort-by foo
╭───┬───╮
│ 0 │ 1 │
│ 1 │ 2 │
╰───┴───╯
```

New behavior:

```nushell
$ [2 1] | sort-by foo
Error: nu::shell::incompatible_path_access

  × Data cannot be accessed with a cell path
   ╭─[entry #29:1:17]
 1 │ [2 1] | sort-by foo
   ·                 ─┬─
   ·                  ╰── int doesn't support cell paths
   ╰────
```

</details>

<details>
<summary><code>sort</code> and <code>sort-by</code> output <code>List</code> instead of <code>ListStream</code> </summary>

This isn't a meaningful change (unless I misunderstand the purpose of ListStream), since `sort` and `sort-by` both need to collect in order to do the sorting anyway, but is user observable.

Old behavior:

```nushell
$ ls | sort | describe -d
╭──────────┬───────────────────╮
│ type     │ stream            │
│ origin   │ nushell           │
│ subtype  │ {record 3 fields} │
│ metadata │ {record 1 field}  │
╰──────────┴───────────────────╯
```

```nushell
$ ls | sort-by name | describe -d
╭──────────┬───────────────────╮
│ type     │ stream            │
│ origin   │ nushell           │
│ subtype  │ {record 3 fields} │
│ metadata │ {record 1 field}  │
╰──────────┴───────────────────╯
```

New behavior:

```nushell
ls | sort | describe -d
╭────────┬─────────────────╮
│ type   │ list            │
│ length │ 22              │
│ values │ [table 22 rows] │
╰────────┴─────────────────╯
```

```nushell
$ ls | sort-by name | describe -d
╭────────┬─────────────────╮
│ type   │ list            │
│ length │ 22              │
│ values │ [table 22 rows] │
╰────────┴─────────────────╯
```

</details>

- `sort` now errors when nothing is piped in (`sort-by` already did this)

# Tests + Formatting

I added lots of unit tests on the new sort implementation to enforce new sort behaviors and prevent regressions.

# After Submitting

See [docs PR](https://github.com/nushell/nushell.github.io/pull/1568), which is ~2/3 finished.

### CODE REVIEW

**Comment by fdncred** [comment]
Thanks for this. So far, so good! I'd like to see the closure parameter built into `sort` or `sort-by` or both so we don't have another command. Also, since you don't have examples yet, it would be good to put an example in the description so other people know how to use it.

**Comment by 132ikl** [comment]
Good idea! Maybe `sort custom` and `sort-by` could both be rolled into `sort` 🤔 I'm not sure what to do with `--ignore-case`, `--values`, and `--natural`. Maybe just ignore them when not relevant? That does seem to be what `sort` already does when passing eg. a list<int> to `sort -i` or `sort -v`.

**Comment by NotTheDr01ds** [comment]
Exactly what I came here (from Discord) to suggest.  `sort-by {closure}` seems pretty good.  And yes, either ignore or error when there's a conflicting arg.

**Comment by fdncred** [comment]
Thanks for the examples, they go a long way to helping us to understand. Looks great!

**Comment by 132ikl** [comment]
Just realized this doesn't entirely close #8322 as it doesn't cover group-by.

**Comment by cablehead** [comment]
Thanks for picking this up! `sort-by {closure}` makes the most sense to me too

**Comment by 132ikl** [comment]
Had to do some major refactoring of `sort_utils` to support both cell paths and closures, and I carried over the error handling idea from `sort custom`, definitely makes the code a lot cleaner and easier to follow (and faster!) :smile: 

I saw the note at the top of `sort_utils` suggesting that it could be moved somewhere else, and would be fine with doing that in this PR, but I'm not familiar enough with the codebase to know where else it might go. Currently the only thing that uses it is `sort-by`. I think I will also address #13008 by splitting up `compare_cell_path` into a general `compare_values` function, and then make `compare_cell_path` just follow the cell path and hand the values to `compare_values`. Then, `sort` can just use `compare_values`. I need to verify more carefully still, but the logic from `sort` seems to be copy pasted from `sort_utils`, so it should be fine to move all logic to `sort_utils`.

I'm not sure if these `sort_value` and `sort_value_in_place` functions are still needed, they aren't used anywhere except their respective tests. I'm not sure why they were introduced, the history is a bit hard to follow. Furthermore, they're a bit more unwieldy to use now that you need to construct a `Vec<Comparator>` instead of just a `Vec<String>`. If there isn't any good reason to keep them around, I can remove them.

**Comment by fdncred** [comment]
I still want to be able to sort-by columns too. Does that still work?

**Comment by 132ikl** [comment]
For interactive use, you can do `sort-by` columns in the same way that you would before, for example `ls | sort-by modified` still works the same way as before, just that `modified` is now parsed as a cell path referring to the modified column instead of a string column name. This doesn't work if you pass a string variable to `sort-by` now, but I'm not sure how common of a use case that would be, and it's just a matter of converting string literals to to cell-path literals or using `into cell-path`.

**Comment by fdncred** [comment]
ok, sounds good so far. thanks.

**Comment by fdncred** [comment]
Do you have time to get the ci green?

**Comment by 132ikl** [comment]
The test failures are intentional since I was doing some test-driven development, so I'll get those passing the next time I can take a look. I'll make sure to address the spellcheck and clippy as well.

**Comment by 132ikl** [comment]
Finally cleared out enough of my massive backlog of tasks, so I should be able to wrap this up finally! I just added a good number of unit tests to define behavior for edge cases and prevent regressions to behavior currently observable on `main`. I also documented the invariants which `sort` (from `sort_utils`, not the `sort` command) provides, with corresponding unit tests to enforce those invariants. Here's the remaining tasks for this PR:

- [ ] Implement functionality for (newly introduced) failing unit tests (`null` sorting to end, reintroduce proper natural sorting)
- [ ] Add `sort-by-key`
- [ ] Add pipeline input to `sort-by` closure (so you can use `$in`)
- [ ] Add examples for new features
- [ ] Update documentation with new functionality and more details on invariants provided by sort commands

**Comment by 132ikl** [comment]
Alright, I think this is ~~finally~~ ready for review!

**Review by IanManske — COMMENTED** [review]
Nice work! I love how everything makes much more sense :sweat_smile: and is consistent.

**Comment by IanManske** [comment]
My only complaint is that closures must be all custom closures or all key closures. It would be nice to automatically detect the comparator type based on the number of arguments the closure has. Although, this might be complicated by default arguments. Wonder if @kubouch or @WindSoilder have any ideas. Either way, this isn't blocking, just something to consider.

**Comment by 132ikl** [comment]
Thank you for the feedback! I'll take a look through it in the next couple days.

Regarding detecting number of closure parameters to determine key closure or closure closure, I proposed the idea [in the Discord](https://discord.com/channels/601130461678272522/615329862395101194/1260958410594324520) and @devyn advised against it:

> Yeah I wouldn't recommend that, I think we try not to make built-in commands have argument handling behavior you couldn't do within Nushell itself

I'm still not sure how I feel about it. It _would_ be better ergonomics, but it's a bit magic-y. It would also require removing `$in.0`/`$in.1` for custom closures, since otherwise zero argument closures would be ambiguous.

&nbsp;
I should note that I _think_ it's possible to rewrite any priority sort as an equivalent multiple sort (using cell paths here, but same would apply for a key closure + a custom closure):
```
sort-by foo bar baz
```
is equivalent to
```
sort-by baz | sort-by bar | sort-by foo
```
with the drawback that `baz` has to be compared for all values regardless of whether values of `bar` or `foo` are equal.

**Comment by IanManske** [comment]
Great points! I would also assume that custom comparators would be the rarer case, so having it behind a flag doesn't seem too bad.

**Comment by fdncred** [comment]
Are we ready to land this then?

**Comment by 132ikl** [comment]
I've got to go through Ian's review still, hopefully sometime today. Other than that, I'm good as long as no one else has any concerns before then.

**Comment by fdncred** [comment]
ok, i'm marking it draft until you're done. then we can try and get it landed.

File: crates/nu-command/src/sort_utils.rs [hunk 1]
```diff
@@ -1,351 +1,273 @@
...
-// This module includes sorting functionality that is useful in sort-by and elsewhere.
-// Eventually it would be nice to find a better home for it; sorting logic is only coupled
-// to commands for historical reasons.
+/// A specification of sort order for `sort_by`.
+///
+/// A closure comparator allows the user to return custom ordering to sort by.
+/// A cell path comparator uses the value referred to by the cell path as the sorting key.
+pub enum Comparator {
+    KeyClosure(Closure, EngineState, Stack),
+    CustomClosure(Closure, EngineState, Stack),
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/sort_utils.rs]: Resolved by 90c79804362f8a93d038c66e7841213e1f26db96

File: crates/nu-command/src/filters/sort.rs [hunk 2]
```diff
@@ -134,233 +135,65 @@ impl Command for Sort {
...
-                let mut vec: Vec<_> = pipe_data.into_iter().collect();
-
-                sort(&mut vec, head, insensitive, natural)?;
+            Value::List { ref vals, .. } => {
+                let mut vec = vals.to_owned();
+
+                // If we have a table specifically, then we want to sort along each column.
+                // Record's PartialOrd impl dictates that columns are compared in alphabetical order,
+                // so we have to explicitly compare by each column.
+                if let Type::Table(cols) = value.get_type() {
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/filters/sort.rs]: Suggestion applied in c09002591f81386f5df09c787480e0f49d924a70

File: crates/nu-command/src/sort_utils.rs [hunk 3]
```diff
@@ -1,351 +1,273 @@
...
-                    folded_right.coerce_into_string(),
-                ) {
-                    (Ok(left), Ok(right)) => compare_str(left, right),
-                    _ => Ordering::Equal,
-                }
-            } else {
-                folded_left
-                    .partial_cmp(&folded_right)
-                    .unwrap_or(Ordering::Equal)
+    for cmp in comparators.iter() {
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/sort_utils.rs]: Resolved by 34a538477

File: crates/nu-command/src/sort_utils.rs [hunk 4]
```diff
@@ -1,351 +1,273 @@
...
+    // allow the comparator function to indicate error
+    // by mutating this option captured by the closure,
+    // since sort_by closure must be infallible
+    let mut compare_err: Option<ShellError> = None;
 
-            if !ascending {
-                vals.reverse();
-            }
+    vec.sort_by(|a, b| {
+        crate::compare_values(a, b, insensitive, natural).unwrap_or_else(|err| {
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/sort_utils.rs]: Resolved by 34a538477

File: crates/nu-command/src/filters/sort_by.rs [hunk 5]
```diff
@@ -79,39 +129,62 @@ impl Command for SortBy {
...
+                        err_message:
+                            "Cannot sort using a value which is not a cell path or closure".into(),
+                        span: val.span(),
+                    })
+                }
+            };
+            comparators.push(cmp);
+        }
+
+        crate::sort_by(&mut vec, comparators, head, insensitive, natural)?;
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/filters/sort_by.rs]: Resolved by 4cb88b530

File: crates/nu-protocol/src/value/mod.rs [hunk 6]
```diff
@@ -96,13 +103,6 @@ pub enum Value {
...
         internal_span: Span,
     },
-    String {
-        val: String,
-        // note: spans are being refactored out of Value
-        // please use .span() instead of matching this span value
-        #[serde(rename = "span")]
-        internal_span: Span,
-    },
     Glob {
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-protocol/src/value/mod.rs]: I had missed this, apparently they actually already are sorted as strings (as in, there was already a non-constant ordering between `String`s and `Glob`s)! I moved `Glob` alongside `String` to make this more obvious, and also allowed them to be sorted insensitively and naturally. Rel: a0c7f0e6f1fd38e76afd92f1e4d44970baae85d7

File: crates/nu-protocol/src/value/mod.rs [hunk 7]
```diff
@@ -2346,16 +2346,16 @@ impl PartialOrd for Value {
...
                 Value::Duration { .. } => Some(Ordering::Greater),
                 Value::Date { .. } => Some(Ordering::Greater),
                 Value::Range { .. } => Some(Ordering::Greater),
-                Value::String { .. } => Some(Ordering::Greater),
                 Value::Glob { .. } => Some(Ordering::Greater),
                 Value::Record { .. } => Some(Ordering::Greater),
                 Value::List { .. } => Some(Ordering::Greater),
                 Value::Closure { .. } => Some(Ordering::Greater),
-                Value::Nothing { .. } => Some(Ordering::Greater),
+                Value::Nothing { .. } => Some(Ordering::Less),
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-protocol/src/value/mod.rs]: Resolved by 8e80419376c94e8ad63da3d674d09dbbe675c790

File: crates/nu-command/src/sort_utils.rs [hunk 8]
```diff
@@ -1,351 +1,273 @@
...
-                Value::test_record(record! {
-                "fruit" => Value::test_string("pear"),
-                "count" => Value::test_int(3),
-                            }),
-            ],)
-        );
+    if natural {
+        alphanumeric_sort::compare_str(left_str, right_str)
+    } else {
+        left_str.partial_cmp(right_str).unwrap_or(Ordering::Equal)
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/sort_utils.rs]: Thanks, missed that

Resolved by 74ffb6bd7

File: crates/nu-command/src/sort_utils.rs [hunk 9]
```diff
@@ -1,351 +1,273 @@
...
-        );
+pub fn compare_values(
+    left: &Value,
+    right: &Value,
+    insensitive: bool,
+    natural: bool,
+) -> Result<Ordering, ShellError> {
+    if should_string_compare(left, right, natural) {
+        let left_str = left.coerce_string()?;
+        let right_str = right.coerce_string()?;
```
[the remark this answers is not in the document]
Comment (132ikl) [reply crates/nu-command/src/sort_utils.rs]: Resolved by 45333b5d9

**Comment by 132ikl** [comment]
I think we're good to go on this @fdncred

**Comment by fdncred** [comment]
ok, just need @IanManske's sign-off since he had a bunch of comments.

**Review by IanManske — APPROVED** [review]
Thanks for the fixes, this looks good :rocket:
